src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,180 @@
+
+    private void addByIndex(Node<T> newNode, int index) {
+        Node<T> prevNode = getNode(index - 1);
+        newNode.next = prevNode.next;
+        prevNode.next = newNode;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        validateIndex(index);
+        return getNode(index).element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        validateIndex(index);
+        Node<T> node = getNode(index);
+        T oldValue = node.element;
+        node.element = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validateIndex(index);
+
+        Node<T> removedNode;
+        if (index == 0) {
+            removedNode = head;
+            head = head.next;
+            if (head == null) {
+                tail = null;
+            }
+        } else {
+            Node<T> prevNode = getNode(index - 1);
+            removedNode = prevNode.next;
+            prevNode.next = removedNode.next;
+            if (prevNode.next == null) {
+                tail = prevNode;
+            }
+        }
+        size--;
+        return removedNode.element;
     }
 
     @Override
     public boolean remove(T object) {
+        if (isEmpty()) {
+            return false;
+        }
+
+        Node<T> current = head;
+        Node<T> prev = null;
+
+        while (current != null) {
+            if ((object == null && current.element == null)
+                    || (object != null && object.equals(current.element))) {
+                unlink(current);
+                size--;
+                return true;
+            }
+            prev = current;
+            current = current.next;
+        }
+
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> getNode(int index) {
+        validateIndex(index);
+        Node<T> current = head;
+        for (int i = 0; i < index; ++i) {
+            current = current.next;
+        }
+        return current;
+    }
+
+    private void validateIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+    }
+
+    private void unlink(Node<T> node) {
+        Node<T> prev = findPreviousNode(node);
+
+        if (prev == null) {
+            head = node.next;
+            if (head == null) {
+                tail = null;
+            }
+        } else {
+            prev.next = node.next;
+            if (prev.next == null) {
+                tail = prev;
+            }
+        }
+    }
+
+    private Node<T> findPreviousNode(Node<T> targetNode) {
```

Comment:
it is too complicated, remove that, and just add `prev` field to the `Node` class, as it was pointed out in the 2 previous reviews, so please be attentive to ALL comments. They must be either fixed as suggested or commented on why not